### PR TITLE
Move dag longest path functions to rustworkx-core. 

### PR DESCRIPTION
**Pull request by henryzou50**
### Attempt to Solve [#1168](https://github.com/Qiskit/rustworkx/issues/1168)

**Summary:**
We are moving the `longest_path` function to `rustworkx-core`. The main implementation will be added to `rustworkx-core/src/longest_path.rs`, utilizing only Rust generics. The existing `src/dag_algo/longest_path.rs` will be updated to serve as the Python interface. Since we are retaining this file, updates to `src/dag_algo/mod.rs`, where the four DAG longest path functions reside, are not necessary.

**Details**
- Added `longest_path.rs` to `rustworkx-core/src/`
- Added test cases to `rustworkx-core/src/longest_path.rs`
- Updated `src/dag_algo/longest_path.rs` to utilize `rustworkx-core/src/longest_path.rs`
- Added release note
- Refined the documentation
- Made slight optimizations

### CODE REVIEW

File: src/dag_algo/longest_path.rs [hunk 1]
```diff
@@ -28,52 +26,22 @@ where
...
-            us.push((length, p_node));
+
+    // Create a new weight function that matches the required signature
+    let edge_cost = |edge_ref: EdgeReference<'_, PyObject>| -> T {
+        let source = edge_ref.source().index();
+        let target = edge_ref.target().index();
+        let weight = edge_ref.weight();
+        match weight_fn(source, target, weight) {
+            Ok(w) => w,
+            Err(_) => T::zero(),
```
Comment (henryzou50) [inline src/dag_algo/longest_path.rs:37]: I don't think this the correct way to handle invalid edge weights, but I am unsure how to fix it. Currently we fail the test: 
```
FAIL: test_nan_not_valid_weight (digraph.test_depth.TestWeightedLongestPath.test_nan_not_valid_weight)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/henryzou/Documents/GitHub/rustworkx/tests/digraph/test_depth.py", line 287, in test_nan_not_valid_weight
    with self.assertRaises(ValueError):
AssertionError: ValueError not raised
```

where in the test the weight function is:

``` python
def weight_fn(*_):
            return float("nan")
 ```

**Review by mtreinish — COMMENTED** [review]
Overall this looks great, thanks for doing this @henryzou50! I left a couple inline comments the only two things are really the return type for the output `Vec` when a path is found and the new module's name. Otherwise I think this is good to go.

File: rustworkx-core/src/longest_path.rs [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+/// let n2 = graph.add_node(());
+/// graph.add_edge(n0, n1, 1);
+/// graph.add_edge(n0, n2, 3);
+/// graph.add_edge(n1, n2, 1);
+///
+/// let weight_fn = |edge: petgraph::graph::EdgeReference<i32>| Ok::<i32, &str>(*edge.weight());
+/// let result = longest_path(&graph, weight_fn).unwrap();
+/// assert_eq!(result, Some((vec![n0.index(), n2.index()], 3)));
+/// ```
+pub fn longest_path<G, F, T, E>(graph: G, mut weight_fn: F) -> Result<Option<(Vec<usize>, T)>, E>
```
[the remark this answers is not in the document]
Comment (henryzou50) [reply rustworkx-core/src/longest_path.rs:66]: Using G::NodeId sounds like a great idea for simplifying user interactions. I will do that.

**Comment by henryzou50** [comment]
@mtreinish thanks for the review and comments! I applied the changes you suggested and it should be ready to go now.

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for the quick update and your first contribution to rustworkx!

**Pull request merged**